Model generation support for SatML

Halbaroth opened the pull request:
This PR includes two things:
- A refactoring of the way AE prints models.
- A support of the model generation for SatML.
I cannot separate them as the support for model generation in SatML has been done after the refactoring in the OptimAE branch.

**IMPORTANT**: This PR has to been merged before OptimAE.

~~I also disable some tests as the printers change a bit but the models (with FunSAT) are still valid. I am preparing another PR rebase on this one that fix printers and test the support of model generation with SatML.~~

iguerNL committed:
Theory: harden 'compute_concrete_model' function

iguerNL committed:
Push the test 'get_case_split_policy() == origin' into Theory.do_case_split

iguerNL committed:
CS: functions return all the env instead of just 'gamma_finite'.

iguerNL committed:
satML: add missing case-split strategies

iguerNL committed:
Fun_sat: refactor the way models are handled/printed

- add a data-structure for models,
- save the strucutre in the SAT's env
- print the models in the Frontend module

iguerNL committed:
satML: implement models generation.

Halbaroth committed:
Make model tests negative

Some tests about models failed in the OptimAE PR.
This commit allows to tag tests in `tests/` with `fail` tag
which means the test is supposed to fail.

Halbaroth committed:
Reindent

bclement-ocp commented:
Thanks for splitting the PR up — I will take a look at this one in the afternoon.

(We could split up further by merging the refactoring first as well — but this is at a size and scope that is manageable on its own now I think)

Halbaroth commented on src/lib/frontend/frontend.ml:
```diff
@@ -246,27 +273,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
       print_status (Sat (d,t)) (Steps.get_steps ());
-      (*if get_model () then SAT.print_model ~header:true (get_fmt_mdl ()) t;*)
-      env, `Sat t, dep
+      print_model env (Some SAT.NoTimeout);
```
We should prevent from printing model here as we did before this PR.

bclement-ocp replied:
Yes, let's undo this change.

Halbaroth replied:
On further consideration, the line is correct. A part of the model refactoring consists in pulling the model printing from the `Models.ml` to the `Frontend.ml`. This code is now correct and will be changed after Steven's PR.

Halbaroth commented on src/lib/frontend/frontend.ml:
```diff
@@ -246,27 +273,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
-      (*if get_model () then SAT.print_model ~header:true (get_fmt_mdl ()) t;*)
-      env, `Sat t, dep
+      print_model env (Some SAT.NoTimeout);
+      env , consistent, dep
```
This line reverts what we did in #789. Still the model printed for the test `777.smt2` is correct. We should investigate.

bclement-ocp replied:
The last generated model is now saved imperatively on the solver using a reference, so this means that when we get the model from the old environment we still find the model of the new environment. Not very fond of this design, but we can revisit it later.

Halbaroth replied:
I would be better to have a fully immutable environment as the API of the SAT solvers hints the environment is immutable. I agree we should fix that in another PR. Probably after the Steven's PR on the frontend.

Halbaroth commented on src/lib/frontend/models.ml:
```diff
@@ -477,55 +484,63 @@ let rec pp_value ppk ppf = function
-      Why3CounterExample.output_constraints fmt props;
-      Format.fprintf fmt "@ ; values"
+let output_concrete_model fmt m =
+  if Options.get_interpretation () then begin
```
We should remove this test. Deciding printing models shouldn't be taken in the pretty printer but in some caller of this function.

bclement-ocp replied:
Yes. Created an issue.

Halbaroth commented on src/lib/frontend/models.mli:
```diff
@@ -30,16 +30,17 @@
+  arrays : ModelMap.t;
+}
+
 (** Print the given counterexample on the given formatter with the
     corresponding format set with Options.get_output_format.
     - functions: the functions of the model;
     - constants: the variables of the model;
     - arrays: (experimental) the arrays of the model.
 *)
```
We have to up to date this comment.

bclement-ocp replied:
Sure. Do you want to do it in this PR, or create an issue to do it later?

Halbaroth replied:
I did it in this PR.

Halbaroth commented on src/lib/frontend/models.mli:
```diff
@@ -30,16 +30,17 @@
-  constants:ModelMap.t ->
-  arrays:ModelMap.t ->
-  unit
+val output_concrete_model : Format.formatter -> t -> unit
```
As we use `Fmt`, we can replace this signature by `t Fmt.t`.

bclement-ocp replied:
Sure. Want to do it?

Halbaroth commented on src/lib/util/printer.ml:
```diff
@@ -369,10 +369,12 @@ let print_status_unknown ?(validity_mode=true) loc
     ("I don't know","unknown","fg_cyan") loc
     time steps goal
 
+(* TODO: The timeout answer doesn't exist in the SMT-LIB standard.
+   We should change this function. *)
```
I leave this comment here to fix it later. We should always print SMT-LIB format answer with SMT-LIB files. For the legacy output format, I have no strong opinion but having more possible answers could be confusing for the final user.

bclement-ocp replied:
It is useful to inform users that an unknown result is due to a timeout being reached and that re-trying with a higher timeout may succeed. Z3 uses a non-standard "timeout" response in that case. The SMT-LIB way of getting this information is to implement `(get-info :reason-unknown)`, but this is out of scope for this PR. I think both "timeout" and "unknown" are acceptable here.

Halbaroth replied:
Let's keep `unknown` here and we will fix this in another PR. I will open an issue and put a link here.

Halbaroth replied:
Issue: #834

Halbaroth commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1902,10 +1845,13 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
     (* all_models_sat_env := None; *)
-    latest_saved_env := None;
-    terminated_normally := false;
+    (* latest_saved_env := None;
+       terminated_normally := false; *)
```
Probably wrong, I will fix it in a commit on this PR.

bclement-ocp replied:
Can you clarify what is probably wrong? These two references no longer exist.

Halbaroth commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1810,10 +1750,12 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
       Debug.bottom classes;
       raise (Unsat d)
-    | Util.Timeout when switch_to_model_gen env -> do_switch_to_model_gen env
+    | Util.Timeout ->
+      (* don't attempt to compute a model if timeout before
+         calling unsat function *)
+      raise (I_dont_know {env; timeout = Assume})
```
This modification makes sense in the native input language. We don't expect a counterexample of a wrong context. But in the SMT-LIB, we should check what's happened.

bclement-ocp replied:
I think this is the same independent of input language. If the comment is correct, for a SMT-LIB problem, if we timeout here it means that we timed out before processing the `(check-sat)` command, which means that we may not have processed all the assertions / definitions. We shouldn't try to produce a model in this case.

Halbaroth replied:
You are right. I had a doubt.

Halbaroth commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -196,9 +192,15 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
     { env with
       unit_facts_cache = ref refs.unit_facts}, guard
 
+  type timeout_reason =
+    | NoTimeout
+    | Assume
+    | ProofSearch
+    | ModelGen
+
```
We should add a documentation for this ADT in this PR.

Halbaroth commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1326,12 +1271,14 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
   let greedy_instantiation env =
     match Options.get_instantiation_heuristic () with
     | INormal ->
+      (* TODO: check if this test still produces a wrong model. *)
```
I added this comment as a remainder.

Halbaroth committed:
Remove SAT solver choices while model generating

As SatML supports model generation, we don't need to select
the appropriate SAT solver while parsing the command line or
the SMT-LIB statement `(set-option :produce-models true)`.

Halbaroth committed:
Fix model output

We should print constraints in models only if the appropriate flag is
used in the command line.

Halbaroth committed:
All the model tests are positive

Halbaroth committed:
Remove deprecated cram tests

Some cram tests are not valid anymore as we don't need to select the SAT
solver Tableaux while generating models.

Halbaroth committed:
Test models with all the SAT solvers

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -489,7 +489,17 @@ let main () =
         if Options.get_interpretation () then
           match State.get partial_model_key st with
-          | Some partial_model -> SAT.get_model partial_model; st
+          | Some partial_model ->
+            begin
+              match SAT.get_model partial_model with
+              | Some (lazy model) ->
+                Models.output_concrete_model
+                  (Options.Output.get_fmt_regular ()) model;
+                st
+              | _ ->
+                (* TODO: is it reachable? *)
+                st
+            end
```
Why do we do this dance? It looks cleaner to call `SAT.get_model` at the time we set the `partial_model_key` (this is OK since the model is lazy).

Halbaroth replied:
It's done.

Halbaroth replied:
Besides, the `partial_model_key` is now an option on a closure of the model. But we will need again to access to the last consistent environment in order to implement `get-info` in `Solving_loop` (Steven is working on it). A better design consists in modifying the signature of `get_model` in the module `Sat_sig`.  The function should return a `Models.t option` instead of `Models.t Lazy.t option`. If you call `get_model`, you want it and the value has to be force in `get_model` directly and keep the last consistent environment in `partial_model_key`. We should rename this key by the way as I will use it for different thing. Maybe `consistent_env_key`?

Halbaroth replied:
As explained offline, this modification will cause a conflict with Steven's PR. I reverted it.

bclement-ocp commented on src/lib/frontend/frontend.ml:
```diff
@@ -246,27 +273,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
       (* In this case, we obviously want to print the status,
          since we exit right after  *)
       print_status (Timeout (Some d)) (Steps.get_steps ());
+      print_model env None;
```
Here we probably don't want to print the model either (or if we do it should be a separate option).

Halbaroth replied:
It works because we check if we want to print models in the function `output_concrete_model`. We will change this while merging Steven's PR on the frontend.

bclement-ocp commented on src/lib/frontend/frontend.ml:
```diff
@@ -246,27 +273,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
+        else (Unknown (d, env))
+      in
+      print_status status (Steps.get_steps ());
+      print_model env (Some timeout);
```
Same as above, probably shouldn't print here.

bclement-ocp commented on src/lib/reasoners/satml_frontend.ml:
```diff
@@ -972,13 +982,72 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
+    else
+      begin
+        (* Beware: models generated on timeout of ProofSearch phase may
+           be incohrent wrt. the ground part of the pb (ie. if delta
```
Probably "may not be coherent" or "may be incoherent"?

Halbaroth replied:
I chose the second one.

bclement-ocp commented on src/lib/reasoners/satml.ml:
```diff
@@ -703,17 +705,13 @@ module Make (Th : Theory.S) : SAT_ML with type th = Th.t = struct
-      with Ex.Inconsistent (expl, _) ->
-        C_theory expl
-    else C_none
+    try
+      let tenv, _terms = Th.do_case_split env.tenv origin in
+      (* TODO: terms not added to matching !!! *)
+      env.tenv <- tenv;
+      C_none
+    with Ex.Inconsistent (expl, _) ->
+      C_theory expl
```
Shouldn't we keep the `if Options.get_case_split_policy () == origin` check here?

bclement-ocp replied:
(Sorry, forgot to remove this comment — the check is now in `Theory.ml`)

bclement-ocp commented on src/lib/reasoners/theory.mli:
```diff
@@ -60,14 +62,7 @@ module type S = sig
-    unit
-
   val reinit_cpt : unit -> unit
-  (** reinitializes the counter to zero *)
```
Should keep the comment?

Halbaroth replied:
I kept it even if it is pointless in my opinion. We should rename the function as we don't care the function resets an internal counter if we don't document the purpose of this counter.

bclement-ocp reviewed (changes requested):
There are a couple minor cleanups to do, and some other things we may want to change later (e.g. the use of a reference for the last env), but I think we should merge with only the minor cleanups for now so that we can move on.

Halbaroth committed:
Update documentation about `output_concrete_model`

I also remove the mention about the different kind of formats used
to print models. Indeed, we use only the SMT-LIB format and
the Why3 format is slightly different but probably outdated.

Halbaroth committed:
Document the timeout_reason ADT

Halbaroth committed:
poetry

Halbaroth committed:
Restore the comment about `reinit_cpt`

Halbaroth committed:
Fix spelling

Halbaroth commented:
> There are a couple minor cleanups to do, and some other things we may want to change later (e.g. the use of a reference for the last env), but I think we should merge with only the minor cleanups for now so that we can move on.

Seems good to mean as we have another PR to fix and test the feature #832

Halbaroth committed:
Remainder to the issue 834

Halbaroth committed:
Simplify get-model processing

Halbaroth committed:
Restoring dump-models option

The dump-models option have been broken by the refactoring in OptimAE.
I restore this feature.

Halbaroth committed:
Use the right post-solve SAT environment for models

Return the appropriate environment in the Frontend module to
retrieve the model with `(get-model)` as we did in the PR #789.

Halbaroth commented:
I think we can merge this PR. The PR #832 fixes almost everything but we have to merge #836 and #835 in order to pass the CI on it.

bclement-ocp commented:
Is #836 really needed? PROXY variables shouldn't end up in models, unless I am mistaken.

If #832 is ready and working we should merge #832 into this PR, then merge both into next at once.

Halbaroth commented:
> Is #836 really needed? PROXY variables shouldn't end up in models, unless I am mistaken.

They do :/ I will check why they leak.

> If #832 is ready and working we should merge #832 into this PR, then merge both into next at once.

Ok

Halbaroth committed:
Print dump models on the channel models

This commit restore the feature of the PR #838 as the
location of code moved after the refactoring of model generation in
OptimAE.

Halbaroth committed:
Prevent printing internal names in models

Symbol names prefixed with a dot or @ shouldn't never be printed
in models as they cannot be declared by the user.

This behaviour is SMT-LIB compliant.

Halbaroth committed:
Promote tests

Halbaroth commented:
The CI will fail because `FunSAT` failed to produce a complete model for the first `check-sat` in `bool1.smt2`. `SatML` manages to produce a complete model in this case but the expected output has to be the some for both solvers. I will fix the issue with `FunSAT` in another PR.

Halbaroth commented:
I forgot to add the support for `String.starts_with` in the `compat` module! This feature is not present in old stdlib. I'll do it right now.

You did it!

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -223,7 +223,7 @@ let main () =
   in
 
-  let partial_model_key: SAT.t option State.key =
+  let partial_model_key: Models.t Lazy.t option State.key =
```
I believe we decided on undoing this change for compatibility with #837

bclement-ocp replied:
This was solved as I was reviewing

bclement-ocp commented on src/lib/frontend/frontend.ml:
```diff
@@ -246,27 +270,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
-      (*if get_model () then SAT.print_model ~header:true (get_fmt_mdl ()) t;*)
-      env, `Sat t, dep
+      print_model env (Some SAT.NoTimeout);
+      env , consistent, dep
```
Shouldn't we still return `` `Sat t`` here?

Halbaroth replied:
Good catch!

bclement-ocp commented on src/lib/frontend/frontend.ml:
```diff
@@ -246,27 +270,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
-      (*if get_model () then SAT.print_model ~header:true (get_fmt_mdl ()) t;*)
-      env, `Unknown t, dep
+      env , `Unsat, dep
+    | SAT.I_dont_know {env; timeout} ->
```
This `env` shadows whereas previously there was no shadowing, I think we need

```suggestion
    | SAT.I_dont_know {env = t; timeout} ->
```

and corresponding changes below.

Halbaroth replied:
I agree this shadowing is misleading but we have to return in this branch the SAT environment fron the caught exception:
```ocaml
    | SAT.I_dont_know {env = t; timeout} ->
      (* TODO: always print Unknown for why3 ? *)
      let status =
        if timeout != NoTimeout then (Timeout (Some d))
        else (Unknown (d, t))
      in
      print_status status (Steps.get_steps ());
      print_model t (Some timeout);
      (* TODO: Is it an appropriate behaviour? *)
      (*       if timeout != NoTimeout then raise Util.Timeout; *)
      env, `Unknown t, dep

```

bclement-ocp replied:
Yeah, the first element of the returned triple ``env, `Unknown t, dep`` is `env` which is shadowed I believe? From what I remember the first element needs to be `env` here for correctness (otherwise the env has additional assertions that are not part of the semantic) but I may be wrong

Halbaroth replied:
Oh you are right! The previous code was wrong because of the shadowing. Now it's fixed. We should add some tests for push and pop, otherwise we cannot detect this kind of mistake easily.

Halbaroth committed:
Revert "Simplify get-model processing"

This reverts commit 9507d632babad381cd29c2a40c997da034cf43ce.

bclement-ocp commented on tests/dune.inc:
Shouldn't this have been excluded from the diff by #830 ?

Halbaroth replied:
I will investigate.

Halbaroth replied:
I will solve this in another PR.

bclement-ocp commented on src/lib/structures/symbols.ml:
```diff
@@ -122,6 +122,12 @@ let is_ac x = match x with
+let is_internal sy =
+  match sy with
+  | Name (hs, _, _) ->
+    Compat.String.starts_with ~prefix:"." (Hstring.view hs)
```
Also `"@"`

bclement-ocp reviewed (changes requested):
A couple very minor things to fix and we should be good to go :tada:

Halbaroth committed:
Review changes

Halbaroth committed:
Test only models with FunSAT

Prevent to test models with other solvers as we cannot
guarantee they will produce exactly the same model.
We have to modify gentest to manage several expected files while
testing models.

Halbaroth commented:
I removed tests for other solvers. I will modify `gentest` script later to manage tests with several expected outputs but I have more urgent stuff to do now and this PR should be merge as soon as possible.

bclement-ocp reviewed (approved):
LGTM :tada: 

Now only check-all-sat and OptimAE itself remain?

Halbaroth commented:
Yes! I'm working on `CheckAllSat`.
